Adds design doc for bundle CRD (#2)
cert-manager/trust: design/20210715.bundle-crd.md

---
title: Bundle CRD
authors:
  - "@joshvanl"
editor: "@joshvanl"
creation-date: 2021-07-15
last-updated: 2021-07-15
status: implemented
---

# Bundle CRD

## Table of Contents

<!-- toc -->
- [Summary](#summary)
- [Motivation](#motivation)
- [Goals](#goals)
  * [Non-Goals](#non-goals)
- [Proposal](#proposal)
  * [Bundle](#bundle)
    + [API](#api)
    + [Custom Cache](#custom-cache)
    + [Upstream CertificateBundle](#upstream-certificatebundle)
  * [Testing](#testing)
  * [Risks and Mitigations](#risks-and-mitigations)
  * [Graduation](#graduation)
      - [Alpha](#alpha)
      - [Beta](#beta)
      - [GA](#ga)
<!-- /toc -->

## Summary

[Bundle](https://github.com/cert-manager/trust/blob/44d8fcaeec7232daf6dd14a1be5a6d0791bc23b8/pkg/apis/trust/v1alpha1/types_bundle.go)
is a cluster scoped CRD designed to distribute trust data across a cluster so
that it is made available by all services that need it. Bundle is intended to be
complimentary to [cert-manager](https://github.com/jetstack/cert-manager) by

> allowing access to trust data that is both known to cert-manager, and external
> to any cert-manager [Issuer](https://cert-manager.io/docs/configuration/).

Reviewer: nitpick: I think this reads like trust data needs to be both known and external to cert-manager

```suggestion
allowing access to trust data that is either known to cert-manager or external
to any cert-manager [Issuer](https://cert-manager.io/docs/configuration/).
```

Services often need multiple CA certificates as part of the same bundle or file.

The upstream
[Kubernetes CertificateSigningRequest resource](https://kubernetes.io/docs/reference/access-authn-authz/certificate-signing-requests/)
does not include any CA field. This means consumers after forced,
[and are intended to](https://kubernetes.io/docs/reference/access-authn-authz/certificate-signing-requests/#kubernetes-signers),
find an out of band method of propagating CA data.

Some users have success by baking in CA data within built container images.
This however requires the assumption that CA data is: known before runtime, long
lived, and all images that require the CA data are built in house.

Projects such as [Istio](https://istio.io/) (as well as Kubernetes itself) has
had success with distributing trust roots, at runtime, by using a
[well-known ConfigMap `istio-ca-root-cert`](https://istio.io/latest/docs/tasks/security/cert-management/plugin-ca-cert/)
that is available within every namespace in a cluster. Bundle is an extension of
this idea- to allow users to use the same approach with a resource that is
expressive enough to meet most needs.

> ## Goals

Reviewer: question: :warning: I think this is an important one

Something that I don't think is mentioned in this document is updating a trust store. It seems like a hard problem, because I think in practice an update would require a cluster operator to roll every deployment which consumed the updated trust store, but equally it's an important thing to have.

Without the ability to update a trust store, we can't:

- rotate an expiring root
- replace/remove an exposed certificate
- easily update publicly-trusted CAs lists (if supported), which are regularly changed

It feels like we need a story on this; do you have any thoughts on it?

Author: >  Something that I don't think is mentioned in this document is updating a trust store.

This is left up to the cluster operator and shouldn't be something that Bundle is concerned with. Bundle will ensure all targets are up to date at all times, but the cluster operator must use either: 1. use moderns apps that watch for changes to trust data files, or 2. use some custom solution like a sidecar or manual rollout.

Reviewer: That's a fair position to take; it might be worth adding something to that effect into the doc

Author: Yes, I agree.

- Create a `Bundle` resource to distribute trust bundles to be made available to
    all services that need them.
- Assist and help investigate
  [upstream](https://github.com/kubernetes/kubernetes/issues/63726) the best
  solution for trust distribution in Kubernetes.
- Enable _most_ users to effectively distribute trust data in their clusters.

### Non-Goals
- Enable _every_ user's use case.
- Compete with upstream development.

## Proposal

### Bundle

A single resource `Bundle` within the `trust.cert-manager.io` group makes up
the trust operator. This resource fetches and appends data from a number of
`sources` in the trust namespace, and distributes them to a `target`.

The trust namespace is the same namespace that the trust operator is deployed
to, which would normally be the `cert-manager` namespace.

While sources currently support `ConfigMap`, `Secret` and `InLine`, it could be
extended to other resources in future, or even remote servers in the aid of

> custom informers.

Author: In two minds about this.

While I absolutely agree this would be a great addition, I feel we run the risk of;
1. The baked in trust bundles becoming out of date meaning the tool has an expiry data if left unmaintained,
2. Open a flood gate of "also add this default trust store" feature creep.

Similarly, target only supports `ConfigMap` but can be extended to other
resources or backends in future.

Each source data has spacing trimmed at the prefix and suffix, and is appended
together with a new line character.

> 

Author: I'll make that clear.

We _don't_ remove any such comments, and specifically have no requirements on the contents of the data.
